programs/manifest/src/state/market.rs

```diff
@@ -0,0 +1,1077 @@
+
+        let mut total_base_atoms_traded: BaseAtoms = BaseAtoms::ZERO;
+        let mut total_quote_atoms_traded: QuoteAtoms = QuoteAtoms::ZERO;
+
+        // Bump the order sequence number even for IOC orders which do not end
+        // up resting.
+        let order_sequence_number: u64 = fixed.order_sequence_number;
+        fixed.order_sequence_number += 1;
+
+        let mut remaining_base_atoms: BaseAtoms = num_base_atoms;
+        while remaining_base_atoms > BaseAtoms::ZERO && current_order_index != NIL {
+            let next_order_index: DataIndex =
+                get_next_candidate_match_index(fixed, dynamic, current_order_index, is_bid);
+
+            let other_order: &RestingOrder =
+                get_helper::<RBNode<RestingOrder>>(dynamic, current_order_index).get_value();
+            trace!("PRE MATCH rem:{remaining_base_atoms} now:{now_slot} order:{other_order:?}");
+
+            // Remove the other order if expired.
+            if other_order.is_expired(now_slot) {
+                let other_is_bid: bool = other_order.get_is_bid();
+
+                let amount_atoms_to_return: u64 = if other_is_bid {
+                    other_order
+                        .get_price()
+                        .checked_quote_for_base(other_order.get_num_base_atoms(), false)?
+                        .as_u64()
+                } else {
+                    other_order.get_num_base_atoms().as_u64()
+                };
+
+                // Global order balances are accounted for on the global accounts, not on the market.
+                if other_order.is_global() {
+                    let global_trade_accounts_opt: &Option<GlobalTradeAccounts> = if other_is_bid {
+                        &global_trade_accounts_opts[1]
+                    } else {
+                        &global_trade_accounts_opts[0]
+                    };
+                    try_to_remove_from_global(&global_trade_accounts_opt, amount_atoms_to_return)?
+                } else {
+                    update_balance_by_trader_index(
+                        dynamic,
+                        other_order.get_trader_index(),
+                        !other_is_bid,
+                        true,
+                        amount_atoms_to_return,
+                    )?;
+                }
+                remove_order_from_tree_and_free(fixed, dynamic, current_order_index, !is_bid)?;
+                current_order_index = next_order_index;
+                continue;
+            }
+
+            // Stop trying to match if price no longer satisfies limit.
+            if (is_bid && other_order.get_price() > price)
+                || (!is_bid && other_order.get_price() < price)
+            {
+                break;
+            }
+
+            // Got a match. First make sure we are allowed to match.
+            assert_can_take(order_type)?;
+
+            if order_type == OrderType::PostOnlySlide {
+                // Post only slide creates a zero spread book.
```

Comment:
add aka a choice market

Reply:
will do in a wrapper. I dont think devs reading code will know what that means though